build/scripts/generate-args.R

```diff
@@ -0,0 +1,262 @@
+    "https://download2.rstudio.org/server/{os_ver}/amd64/rstudio-server-{rstudio_version}-amd64.deb"
+  ) |>
+    stringr::str_replace_all("\\+", "-") |>
+    httr2::request() |>
+    httr2::req_error(is_error = \(...) FALSE) |>
+    httr2::req_perform() |>
+    httr2::resp_is_error() |>
+    isFALSE()
+
+  is_available
+}
+
+
+#' Get the latest CTAN URL for a given date
+#' @param date A [Date] class vector.
+#' If `NA`, the "latest" URL will be returned.
+#' @return A character of CTAN URL.
+#' @examples
+#' latest_ctan_url(as.Date(c("2023-10-30", NA)))
+latest_ctan_url <- function(date) {
+  .url <- dplyr::if_else(
+    is.na(date), "https://mirror.ctan.org/systems/texlive/tlnet",
+    stringr::str_c("https://www.texlive.info/tlnet-archive/", format(date, "%Y/%m/%d"), "/tlnet")
+  )
+
+  .url
+}
+
+
+#' Get the latest version from Git remote tags
+#' @param remote_repo A single character of Git remote repository URL.
+#' @return A character of the latest version.
+#' @examples
+#' latest_version_of_git_repo("https://github.com/OSGeo/PROJ.git")
+latest_version_of_git_repo <- function(remote_repo) {
+  gert::git_remote_ls(remote = remote_repo) |>
+    dplyr::pull(ref) |>
+    stringr::str_subset(r"(^refs/tags/v?(\d+\.){2}\d+$)") |>
+    stringr::str_extract(r"((\d+\.)*\d+$)") |>
+    package_version() |>
+    sort() |>
+    utils::tail(1) |>
+    as.character()
+}
+
+
+#' Paste each element of vectors in a cartesian product
+#' @param ... Dynamic dots. Character vectors to paste.
+#' @return A character vector.
+#' @examples
+#' outer_paste(c("a", "b"), "-", c("c", "d", "e"))
+outer_paste <- function(...) {
+  .paste <- function(x, y) {
+    outer(x, y, stringr::str_c) |>
+      c()
+  }
+
+  out <- rlang::list2(...) |>
+    purrr::reduce(.paste)
+
+  out
+}
+
+
+rocker_versioned_args <- function(
+    ...,
+    r_versions_file = "build/variables/r-versions.tsv",
+    ubuntu_lts_versions_file = "build/variables/ubuntu-lts-versions.tsv",
+    rstudio_versions_file = "build/variables/rstudio-versions.tsv") {
+  df_all <- readr::read_tsv(r_versions_file, show_col_types = FALSE) |>
+    dplyr::arrange(as.numeric_version(r_version)) |>
+    dplyr::mutate(
+      r_minor_version = stringr::str_extract(r_version, r"(^\d+\.\d+)") |>
+        as.numeric_version(),
+      r_major_version = stringr::str_extract(r_version, r"(^\d+)") |>
+        as.numeric_version(),
+    ) |>
+    dplyr::mutate(
+      r_minor_latest = dplyr::if_else(dplyr::row_number() == dplyr::n(), TRUE, FALSE),
+      .by = r_minor_version
+    ) |>
+    dplyr::mutate(
+      r_major_latest = dplyr::if_else(dplyr::row_number() == dplyr::n(), TRUE, FALSE),
+      .by = r_major_version
+    ) |>
+    dplyr::select(!c(r_minor_version, r_major_version)) |>
+    # Supports the latest two patch versions and the latest two minor versions.
+    dplyr::filter(
+      r_minor_latest | dplyr::lead(r_major_latest, default = FALSE)
+    ) |>
+    dplyr::slice_tail(n = 3) |>
```

Comment:
It should be changed so that it is not updated except for the two most recent versions.
Perhaps we need to change the file deletion to be determined by the R version rather than by the `rm` command as it is currently done for Dockerfiles.

Reply:
agreed